notifications_utils/template.py

```diff
@@ -516,45 +514,21 @@ def subject(self):
     def placeholders(self):
         return super().placeholders | get_placeholders(self.contact_block)
 
-    @property
-    def values_with_default_optional_address_lines(self):
-        keys = Columns.from_keys(
-            set(self.values.keys()) | {
-                'address line 3',
-                'address line 4',
-                'address line 5',
-                'address line 6',
-            }
-        ).keys()
-
-        return {
-            key: Columns(self.values).get(key) or ''
-            for key in keys
-        }
-
     @property
     def _address_block(self):
-        return Take(Field(
+        from notifications_utils.postal_address import PostalAddress
+
+        postal_address = PostalAddress.from_personalisation(Columns(self.values))
+
+        if postal_address.has_enough_lines and not postal_address.has_too_many_lines:
+            return postal_address.normalised_lines
+
+        return Field(
             self.address_block,
-            (
-                self.values_with_default_optional_address_lines
-                if all(Columns(self.values).get(key) for key in {
-                    'address line 1',
-                    'address line 2',
-                    'postcode',
-                }) else self.values
-            ),
+            self.values,
             html='escape',
-            with_brackets=False
-        )).then(
-            strip_pipes
```

Comment:
are we still stripping pipes somewhere after those changes? I couldn't find where.

Reply:
Deliberately removing it, explanation in 78d505f966487103d5350bc07dbc998fe91b28cd:

> **Stop stripping pipes from addresses**
>
> This is a holdover from when we used to give DVLA a pipe-delimited file of data, rather than a PDF. Because the data was pipe-delimited the fields themselves couldn’t have pipes in.
>
> Removing this from the address block will let us make this code simpler. Later we can remove it from the other places too.

Reply:
Ahhh, sweet, thanks a lot for pointing me to the commit!